### PR TITLE
Add object mode ReadableStream support

### PR DESCRIPTION
**Pull request by ritch**
This PR adds support for returning object mode `ReadableStream`s.

connect to strongloop/strong-remoting#214

Using HTTP `ReadableStream`s are piped to the `res` using https://github.com/dominictarr/mux-demux. This allows for `stream.on('error')`s to be sent using the multiplexed NL separated JSON protocol.

I would still like to add the same exact capability using multipart responses.

/to @bajtos 
/cc @raymondfeng @kraman @rmg (noticed you had used mux-demux)

### CODE REVIEW

File: lib/http-context.js [hunk 1]
```diff
@@ -304,6 +306,31 @@ HttpContext.prototype.invoke = function(scope, method, fn, isCtor) {
...
+        res.setHeader('Content-Type',
+          streamDesc.contentType || 'application/json; boundary=NL');
+        res.setHeader('Transfer-Encoding', 'chunked');
+
+        var stream = results[streamDesc.arg];
+
+        if (streamDesc.type === 'ReadableStream') {
+          stream.pipe(JSONStream.stringify(false)).pipe(res);
+        } else {
+          fn(new Error('unsupported stream type'));
```
Comment (bajtos) [inline lib/http-context.js]: Please include the actual type value in the error message. If you think it's a sensitive info that should not be returned to the client, then print a debug log. The goal is to simplify troubleshooting.

File: lib/http-context.js [hunk 2]
```diff
@@ -304,6 +306,31 @@ HttpContext.prototype.invoke = function(scope, method, fn, isCtor) {
...
+        }
+      } else {
+        fn(new Error('unsupported stream description'));
```
Comment (bajtos) [inline lib/http-context.js]: Please provide more information to help the user to understand better what they did wrong and how to fix the problem.

Example: `Unsupported stream descriptor, only descriptors with property "json:true" are supported.`.

File: lib/http-context.js [hunk 3]
```diff
@@ -379,6 +406,11 @@ function toXML(input) {
  */
 
 HttpContext.prototype.done = function() {
+  if (this.method.stream) {
+    // the stream has already been setup
+    return;
+  }
+
```
Comment (bajtos) [inline lib/http-context.js]: Is there any particular reason why you cannot handle streams in this `done` method? To me, it seems more logical to handle all different response arguments (like body stream) in a single place.

File: lib/http-context.js [hunk 4]
```diff
@@ -304,6 +306,31 @@ HttpContext.prototype.invoke = function(scope, method, fn, isCtor) {
   var pipe = http && http.pipe;
   var pipeDest = pipe && pipe.dest;
```
Comment (bajtos) [inline lib/http-context.js]: I am proposing to deprecate `http.pipe.dest` in favour of this new "returns stream" API. Thoughts?

File: lib/http-invocation.js [hunk 5]
```diff
@@ -223,14 +228,30 @@ HttpInvocation.prototype.invoke = function(callback) {
...
+  var streamDesc = method.sharedMethod.stream;
+  var methodReturnsStream = !!streamDesc;
+  var streamResult;
+
+  if (methodReturnsStream) {
+    if (streamDesc.type === 'ReadableStream' && streamDesc.json) {
+      streamResult = request(this.req).pipe(JSONStream.parse());
+      callback(null, streamResult);
+    } else {
+      callback(new Error('unsupported stream type'));
```
Comment (bajtos) [inline lib/http-invocation.js]: Please add more info to the error message, make it easier to troubleshoot the problem.

File: test/streams.js [hunk 6]
```diff
@@ -45,3 +47,66 @@ describe('strong-remoting', function() {
...
+      var buf = [];
+      var n = 5;
+      while (n--) {
+        buf.push({n: n});
+      }
+      rs._read = function() {
+        setTimeout(function() {
+          rs.push(buf.shift() || null);
+        }, 1);
+      };
```
Comment (bajtos) [inline test/streams.js]: How about using even-stream's [readArray](https://www.npmjs.com/package/event-stream#readarray-array)? Unless the 1ms timeout is important for your test.

File: test/streams.js [hunk 7]
```diff
@@ -45,3 +47,66 @@ describe('strong-remoting', function() {
...
+    remotes.invoke('StreamClass.createStream', [], function(err, stream) {
+      assert(stream.readable, 'must be a ReadableStream');
+      var count = 0;
+      stream.on('data', function(d) {
+        count++;
+
+        if (count === 5) {
+          done();
+        }
+      });
```
Comment (bajtos) [inline test/streams.js]: Perhaps use [writeArray](https://www.npmjs.com/package/event-stream#writearray-callback) and assert on the array content?

File: lib/http-context.js (hunk 4 above)
Comment (ritch) [inline lib/http-context.js]: Ya... I'll add an issue

File: lib/http-context.js (hunk 3 above)
Comment (ritch) [inline lib/http-context.js]: We need to handle errors and I can't reference the error here (unless there is a `ctx.error` property??).

File: test/streams.js (hunk 6 above)
Comment (ritch) [inline test/streams.js]: Good idea. The `setTimeout` is to add some delay, but I don't think its strictly necessary.

File: test/streams.js (hunk 7 above)
Comment (ritch) [inline test/streams.js]: Good idea... will do!

**Comment by rmg** [comment]
Missing `JSONStream` dependency.. also, please consider switching to `jsonstream` instead - it's the canonical version of the module. Sadly, since OS X uses a case-folding filesystem, it can't actually tell the difference between them :-(

File: lib/http-context.js [hunk 8]
```diff
@@ -38,6 +41,8 @@ function HttpContext(req, res, method, options) {
   this.args = this.buildArgs(method);
   this.methodString = method.stringName;
   this.supportedTypes = this.options.supportedTypes || DEFAULT_SUPPORTED_TYPES;
+  var streamDesc = method.stream;
+  var methodReturnsStream = !!streamDesc;
```
Comment (bajtos) [inline lib/http-context.js]: IIUC, the stream mode is turned on by marking a return argument as `type: 'stream'`. This code seems to check the method metadata instead?

File: test/streams.js [hunk 9]
```diff
@@ -45,3 +48,87 @@ describe('strong-remoting', function() {
...
+describe('a function returning a ReadableStream', function() {
+  var Readable = require('stream').Readable;
+  var remotes = RemoteObjects.create();
+  var streamClass;
+  var server;
+
+  before(function(done) {
+    var test = this;
+    // NOTE: Date object will not be coerced back into a Date
+    var data = test.data = [{foo: 'bar'}, 'bat', false, 0, null];
```
Comment (bajtos) [inline test/streams.js]: I find the comment confusing, because I don't see any Date object in the data?

File: test/streams.js [hunk 10]
```diff
@@ -45,3 +48,87 @@ describe('strong-remoting', function() {
...
+
+    streamClass = new SharedClass('StreamClass', StreamClass);
+
+    streamClass.defineMethod('createStream', {
+      isStatic: true,
+      fn: StreamClass.createStream,
+      returns: [{
+        arg: 'stream',
+        type: 'ReadableStream',
+        json: true
```
Comment (bajtos) [inline test/streams.js]: Should we specify `root:true` as well? Or is that implied by `type:ReadableStream`? Could you please use a different arg name in the test, just to make sure that we don't accidentally match it when deciding whether the argument is of a stream type?

**Comment by bajtos** [comment]
@ritch do you expect any non-javascript consumers of this new stream API? If you do, then could you please add a unit-test showing how to read and parse the stream using plain REST?

The Angular code generator will not support this new stream type OOTB, do you have any plans for that SDK? Should it ignore (skip) any methods that return a stream?

File: test/streams.js (hunk 10 above)
Comment (bajtos) [inline test/streams.js]: I mean use for example `arg: 'result'`' instead of `arg: 'stream'`.

File: lib/shared-method.js [hunk 11]
```diff
@@ -82,6 +82,7 @@ function SharedMethod(fn, name, sc, options) {
   this.documented = (options.documented || fn.documented) !== false;
   this.http = options.http || fn.http || {};
   this.rest = options.rest || fn.rest || {};
+  this.authorization = options.authorization;
```
Comment (bajtos) [inline lib/shared-method.js]: This seems like an unrelated change?

File: lib/shared-method.js [hunk 12]
```diff
@@ -109,6 +110,13 @@ function SharedMethod(fn, name, sc, options) {
   }
   this.returns.forEach(normalizeArgumentDescriptor);
 
+  var firstReturns = this.returns[0];
+  var streamTypes = ['ReadableStream', 'WriteableStream', 'DuplexStream'];
+
+  if (firstReturns && firstReturns.type && streamTypes.indexOf(firstReturns.type) > -1) {
+    this.stream = firstReturns;
```
Comment (bajtos) [inline lib/shared-method.js]: I'd like to eventually add support for streams to input arguments too. Would you mind using a different property name to distinguish between "accepts" and "returns" streams? E.g. "streamResult" or "streamReturn"

File: lib/http-context.js (hunk 8 above)
Comment (bajtos) [inline lib/http-context.js]: Never mind, I see that `method.stream` is set in shared-method.js. I'll delete these two comments.

File: test/streams.js (hunk 9 above)
Comment (ritch) [inline test/streams.js]: The test would fail with a `Date`

File: lib/shared-method.js (hunk 11 above)
Comment (ritch) [inline lib/shared-method.js]: Hmmm... that is odd...?

File: lib/http-invocation.js [hunk 13]
```diff
@@ -223,14 +229,37 @@ HttpInvocation.prototype.invoke = function(callback) {
   if (!this.req) {
     this.createRequest();
   }
-  request(this.req, function(err, res, body) {
-    if (err instanceof SyntaxError) {
-      if (res.status === 204) err = null;
+  var method = this.method;
+  var streamDesc = method.sharedMethod.stream;
+  var methodReturnsStream = !!streamDesc;
+  // var streamResult;
```
Comment (bajtos) [inline lib/http-invocation.js]: please remove commented-out code.

File: lib/http-invocation.js [hunk 14]
```diff
@@ -223,14 +229,37 @@ HttpInvocation.prototype.invoke = function(callback) {
...
+  if (methodReturnsStream) {
+    if (streamDesc.type === 'ReadableStream' && streamDesc.json) {
+      var mdm = new MuxDemux();
+      mdm.on('connection', function(stream) {
+        callback(null, stream);
+      });
+
+      request(this.req).pipe(mdm);
+
+      // streamResult = request(this.req).pipe(JSONStream.parse());
```
Comment (bajtos) [inline lib/http-invocation.js]: please remove commented-out code.

File: lib/http-invocation.js [hunk 15]
```diff
@@ -223,14 +229,37 @@ HttpInvocation.prototype.invoke = function(callback) {
...
+    } else {
+      callback(new Error('unsupported stream type'));
     }
-    if (err) return callback(err);
-    self.res = res;
-    self.transformResponse(res, body, callback);
-  });
+
+    return;
+  } else {
```
Comment (bajtos) [inline lib/http-invocation.js]: It may be my personal preference, but I would omit unnecessary `else` block after `return` and thus keep the amount of whitespace changes lower.

File: test/streams.js (hunk 9 above)
Comment (bajtos) [inline test/streams.js]: I see, could you please expand the comment to explain that the test data includes different object types, but that Date is intentionally excluded?

It would be good to document this limitation somewhere, or even emit a warning.

**Comment by bajtos** [comment]
@ritch I am assigning this back to you, please let me know when the code is ready for the final review.

File: test/streams.js (hunk 10 above)
Comment (ritch) [inline test/streams.js]: > Should we specify root:true as well? Or is that implied by type:ReadableStream

That is implied by `type: ReadableStream`

Comment (ritch) [inline test/streams.js]: Changed the arg to `result`

File: lib/http-invocation.js (hunk 13 above)
Comment (ritch) [inline lib/http-invocation.js]: removed

File: lib/http-invocation.js (hunk 15 above)
Comment (ritch) [inline lib/http-invocation.js]: omitted

File: lib/shared-method.js (hunk 11 above)
Comment (ritch) [inline lib/shared-method.js]: removed

File: lib/shared-method.js (hunk 12 above)
Comment (ritch) [inline lib/shared-method.js]: renamed to `streams` it includes (right now) a `returns` property for the first argument in the callback that is a stream

File: lib/http-invocation.js (hunk 14 above)
Comment (ritch) [inline lib/http-invocation.js]: removed

**Comment by ritch** [comment]
This should be ready to merge @bajtos

**Comment by bajtos** [comment]
@ritch the CI is red (both Travis and Jenkins), PTAL.

File: lib/http-context.js [hunk 16]
```diff
@@ -374,15 +405,97 @@ function toXML(input) {
...
+          out.end();
+        });
+        // TODO(ritch) support multi-part streams
+      } else {
+        cb(new Error('unsupported stream type: ' + returnStreamDesc.type));
+      }
+    } else {
+      cb(new Error('Unsupported stream descriptor, only descriptors ' +
+        'with property "json:true" are supported'));
+    }
```
Comment (bajtos) [inline lib/http-context.js]: Nitpick: I would personally move the contents of the block `if(methodReturnsStream)` to a new method, e.g. `HttpContext.prototype._streamResponse`. The purpose is to keep the `done` method reasonably short.

**Comment by ritch** [comment]
test please

**Comment by bajtos** [comment]
@ritch It seems to me that the commit "Merge and fix iojs failures" is bringing in a bunch of changes already landed on the master branch. Would you mind rebasing this PR on the current master, so that 1) the history is linear 2) the diff on github is not polluted with changes that were already reviewed and landed?

**Comment by ritch** [comment]
@bajtos I'm not going to rewrite the history. It takes forever and is error prone (since the commits conflict on the package deps and dev deps.

This PR has gone on way longer than is practical. Just look at the diff instead of the commits.

**Comment by ritch** [comment]
@bajtos ...actually let me show you the pain I am going through when rebasing and perhaps you can either see why I am refusing... or you might be able to see what I am doing incorrectly and prevent future pain.

Here is what I am doing:

```
git fetch # ensure we have latest from master
git rebase master -i # so we can squash the commits

# outputs...
pick 6d9ba85 Add object mode ReadableStream support
pick 4038439 Add support for Event Source streams

# I change the second "pick" to "s" to squash it into the previous commit
# note the merge commit doesn't show up...

# but git is unable to perform the rebase:

could not apply 6d9ba85... Add object mode ReadableStream support
```

If I don't try to squash the commit, the rebase outputs:

```
First, rewinding head to replay your work on top of it...
Applying: Add object mode ReadableStream support
Using index info to reconstruct a base tree...
M	lib/http-context.js
M	lib/rest-adapter.js
M	lib/shared-method.js
M	package.json
M	test/streams.js
Falling back to patching base and 3-way merge...
Auto-merging test/streams.js
Auto-merging package.json
**CONFLICT (content): Merge conflict in package.json**
Auto-merging lib/shared-method.js
Auto-merging lib/rest-adapter.js
Auto-merging lib/http-context.js
**CONFLICT (content): Merge conflict in lib/http-context.js**
Failed to merge in the changes.
Patch failed at 0001 Add object mode ReadableStream support
```

Which I can resolve manually. Its just pretty tedious and easy to merge incorrectly since the diff looks like this:

```
  "dependencies": {
<<<<<<< HEAD
    "express": "4.x",
    "body-parser": "^1.12.4",
    "debug": "^2.2.0",
    "eventemitter2": "^0.4.14",
    "cors": "^2.6.0",
    "jayson": "^1.2.0",
    "js2xmlparser": "^0.1.9",
    "async": "^0.9.0",
    "traverse": "^0.6.6",
    "request": "^2.55.0",
    "qs": "^2.4.2",
    "inflection": "^1.7.1",
    "xml2js": "^0.4.8"
  },
  "devDependencies": {
    "bluebird": "^2.9.25",
    "browserify": "~10.2.0",
    "chai": "^2.3.0",
=======
    "async": "^0.9.0",
    "body-parser": "^1.7.0",
    "cors": "^2.5.2",
    "debug": "^2.0.0",
    "eventemitter2": "^0.4.14",
    "express": "4.x",
    "inflection": "^1.4.2",
    "jayson": "^1.1.1",
    "js2xmlparser": "^0.1.3",
    "mux-demux": "^3.7.9",
    "qs": "^2.2.3",
    "request": "^2.42.0",
    "traverse": "^0.6.6",
    "xml2js": "^0.4.4"
  },
  "devDependencies": {
    "bluebird": "^2.9.6",
    "browserify": "~5.11.1",
    "chai": "^1.10.0",
    "event-stream": "^3.3.1",
>>>>>>> Add object mode ReadableStream support
```

Anyways... I don't need help rebasing... I can get all of that to work. Also I am not against rebasing in general or rewriting history. I'm just tired of re-merging the same changes over and over again only to have missed a version number, have that fail the build on ci and then spend time finding that mistake.

So if you have any way to make this a bit easier or less error prone I would love to know.

**Comment by bajtos** [comment]
> So if you have any way to make this a bit easier or less error prone I would love to know.

I am afraid I don't have any better solution, I think this is the inherent cost of long-living feature branches :(

> Just look at the diff instead of the commits. 

That's what I did. The diff on github includes many unrelated changes, so it's rather difficult to distinguish signal from noise. Fortunately git client shows shorter diff, I'll review using that way.

File: package.json [hunk 17]
```diff
@@ -13,24 +13,28 @@
...
-    "qs": "^2.4.2",
-    "inflection": "^1.7.1",
-    "xml2js": "^0.4.8"
+    "xml2js": "^0.4.4"
   },
   "devDependencies": {
-    "bluebird": "^2.9.25",
-    "browserify": "~10.2.0",
-    "chai": "^2.3.0",
+    "bluebird": "^2.9.6",
```
Comment (bajtos) [inline package.json]: AFAICT, you are downgrading the min bluebird version from 2.9.25 to 2.9.6, is that intentional?

File: package.json [hunk 18]
```diff
@@ -13,24 +13,28 @@
...
+    "browserify": "~5.11.1",
```
Comment (bajtos) [inline package.json]: downgrade from ~10.2.0 to ~5.11.1, is it intentional?

**Comment by bajtos** [comment]
Your package.json changes are downgrading certain dependencies, I pointed out two cases in my comments above. Could you please double-check other deps too?

**Comment by bajtos** [comment]
The code looks good otherwise, merge at will.

**Comment by rmg** [comment]
The rebase conflicts in `lib/http-context.js` are because of the merging of master into the branch instead of frequent rebasing to keep up.

The rebase conflicts in `package.json` are because master doesn't list the dependencies in alphabetical order but your commit that added 2 dependencies re-ordered them. This is where `git add -p` is your friend, since it allows you to commit the re-ordering as a separate commit from adding the deps.

As for re-writing history... A PR is a patch set proposal, and rewriting its history is more akin to working with an editor and revising a manuscript before sending it to the printers where it _does_ become part of the canonical history. This is where rebasing is a good thing.

File: package.json (hunk 17 above)
Comment (rmg) [inline package.json]: Downgraded because the merge from master discarded the change from master.

File: package.json (hunk 18 above)
Comment (rmg) [inline package.json]: Downgraded because the merge from master discarded the change from master.

**Comment by rmg** [comment]
The numerous dependency downgrades are the result of the merge commit discarding the changes to package.json that were made to master since this branch started.

**Comment by bajtos** [comment]
It looks to me like not rewriting the history is even more error prone that rewriting it.

**Comment by ritch** [comment]
So I guess rewriting history itself is not the issue. Its manual merging. Merging is error prone and tedious. It sounds like the only way to avoid having to manually merge files is to rebase often (thanks for pointing that out @rmg).

**Comment by ritch** [comment]
test please
